### PR TITLE
JIT: update debug dumping for loop hoisting

### PR DESCRIPTION
**Pull request by AndyAyersMS**
I kept adding stuff like this whenever I wanted to understand why hosting was
doing what it was doing, so figured it likely had lasting value.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

Sample jit dump fragment
```
BB02 [0001]  2       BB01,BB04             4     0 [006..00A)-> BB04 ( cond )                     i Loop Loop0 bwd bwd-target 
BB03 [0002]  2       BB02,BB03            16     1 [00A..01B)-> BB03 ( cond )                     i Loop Loop0 bwd bwd-target align 
BB04 [0004]  2       BB02,BB03             4     0 [01B..024)-> BB02 ( cond )                     i bwd 

Loop Nest L00
optHoistLoopCode for loop L00 <BB02..BB04>:
  Loop body does not contain a call
  Loop has single exit

  USEDEF  (5)={V03 V00 V01 V04 V02}
  INOUT   (4)={V03 V00 V01 V02}
  LOOPVARS(4)={V03 V00 V01 V02}
  Only considering hosting in blocks that dominate exit block BB04

    optHoistLoopBlocks BB02 (weight=  4   ) of loop L00 <BB02..BB04>, firstBlock is true
      [000012] not hoistable: variant child
      [000043] not hoistable: variant child

    optHoistLoopBlocks BB04 (weight=  4   ) of loop L00 <BB02..BB04>, firstBlock is false
      [000036] not hoistable: variant child
      [000009] not hoistable: variant child
```
Note we skip hosting from the inner loop block BB03 because it is conditionally executed. Added todos to the code for this (echoing some of what's over in #35735).

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -5688,6 +5691,7 @@ void Compiler::optHoistLoopCode()
     {
         if (optLoopTable[lnum].lpFlags & LPFLG_REMOVED)
         {
+            JITDUMP("\nLoop L%02u was removed\n", lnum);
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:5694]: We now have `FMT_LP`

File: src/coreclr/jit/optimizer.cpp [hunk 2]
```diff
@@ -5746,7 +5750,7 @@ void Compiler::optHoistLoopCode()
 void Compiler::optHoistLoopNest(unsigned lnum, LoopHoistContext* hoistCtxt)
 {
     // Do this loop, then recursively do all nested loops.
-    CLANG_FORMAT_COMMENT_ANCHOR;
+    JITDUMP("\n%s L%02u\n", optLoopTable[lnum].lpParent == BasicBlock::NOT_IN_LOOP ? "Loop Nest" : "Nested Loop", lnum);
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:5753]: same

File: src/coreclr/jit/optimizer.cpp [hunk 3]
```diff
@@ -5933,10 +5943,23 @@ void Compiler::optHoistThisLoop(unsigned lnum, LoopHoistContext* hoistCtxt)
...
+    //
+    // In particular if we have a loop nest, when scanning the outer loop we should consider hoisting from blocks
+    // in enclosed loops. However, this is likely to scale poorly, and we really should instead start
+    // hoisting inner to outer.
+    //
     ArrayStack<BasicBlock*> defExec(getAllocatorLoopHoist());
     if (pLoopDsc->lpFlags & LPFLG_ONE_EXIT)
     {
         assert(pLoopDsc->lpExit != nullptr);
+        JITDUMP("  Only considering hosting in blocks that dominate exit block " FMT_BB "\n", pLoopDsc->lpExit->bbNum);
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:5962]: typo: hosting

File: src/coreclr/jit/optimizer.cpp [hunk 4]
```diff
@@ -5947,12 +5970,16 @@ void Compiler::optHoistThisLoop(unsigned lnum, LoopHoistContext* hoistCtxt)
         // If we didn't reach the entry block, give up and *just* push the entry block.
         if (cur != pLoopDsc->lpEntry)
         {
+            JITDUMP("  -- odd, we didn't reach entry from exit via dominators. Only considering hosting in entry "
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:5973]: typo: hosting

File: src/coreclr/jit/optimizer.cpp [hunk 5]
```diff
@@ -5947,12 +5970,16 @@ void Compiler::optHoistThisLoop(unsigned lnum, LoopHoistContext* hoistCtxt)
...
+                    "block " FMT_BB "\n",
+                    pLoopDsc->lpEntry->bbNum);
             defExec.Reset();
         }
         defExec.Push(pLoopDsc->lpEntry);
     }
     else // More than one exit
     {
+        JITDUMP("  only considering hosting in entry block " FMT_BB "\n", pLoopDsc->lpEntry->bbNum);
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:5982]: typo: hosting

File: src/coreclr/jit/optimizer.cpp [hunk 6]
```diff
@@ -6032,6 +6059,8 @@ bool Compiler::optIsProfitableToHoistableTree(GenTree* tree, unsigned lnum)
         // Don't hoist expressions that are not heavy: tree->GetCostEx() < (2*IND_COST_EX)
         if (tree->GetCostEx() < (2 * IND_COST_EX))
         {
+            JITDUMP("    tree cost too low: %d < %d (lvc %u >= arc %u)\n", tree->GetCostEx(), 2 * IND_COST_EX,
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:6062]: IMO, spelling out the variables here would be better. (same below)

File: src/coreclr/jit/optimizer.cpp [hunk 7]
```diff
@@ -6415,6 +6475,7 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
                 if (!child.m_invariant)
                 {
                     treeIsInvariant = false;
+                    INDEBUG(failReason = "variant child";);
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:6478]: nit: I presume you only need one of the semicolons (here and below)

**Review by BruceForstall — CHANGES_REQUESTED** [review]
Thanks for doing this!

**Review by BruceForstall — APPROVED** [review]

**Comment by EgorBo** [comment]
Oh, I was going to use "Loop Hoisting" as an example of a jit area that needs more diagnostic info in my write-up 😐

**Pull request merged**